File: src/lib/anki/tag-template.ts
export interface TagContext {
  series: string;
  volume: number;
  page: number;
  date?: Date;
}

const TAG_VARIABLES = ["{series}", "{volume}", "{page}", "{date}"] as const;

export function expandTags(tags: string[], ctx: TagContext): string[] {
  const date = ctx.date ?? new Date();
  const iso = date.toISOString().slice(0, 10);
  const safe = (s: string) => s.replace(/\s+/g, "_");

  const replacements: Record<(typeof TAG_VARIABLES)[number], string> = {
    "{series}": safe(ctx.series),
    "{volume}": String(ctx.volume),
    "{page}": String(ctx.page),
    "{date}": iso,
  };

  return tags
    .map((tag) =>
      TAG_VARIABLES.reduce(
        (acc, key) => acc.split(key).join(replacements[key]),
        tag,
      ),
    )
    .map((tag) => tag.trim())
    .filter((tag) => tag.length > 0);
}

export function listTagVariables(): readonly string[] {
  return TAG_VARIABLES;
}

File: android/app/src/main/java/com/mangashelf/reader/flashcards/data/TagTemplate.kt
package com.mangashelf.reader.flashcards.data

/**
 * Port of the web `src/lib/anki/tag-template.ts` `expandTags()`: substitutes the
 * `{series}` / `{volume}` / `{page}` / `{date}` placeholders in mining tags, collapses whitespace
 * in the series name to underscores, trims, and drops empties. Unknown placeholders are left
 * untouched. `date` is supplied by the caller (ISO `yyyy-MM-dd`) so this stays pure/testable.
 */
object TagTemplate {
    data class Ctx(val series: String, val volume: Int, val page: Int, val date: String)

    private val VARIABLES = listOf("{series}", "{volume}", "{page}", "{date}")
    private val WHITESPACE = Regex("\\s+")

    fun expand(tags: List<String>, ctx: Ctx): List<String> {
        val replacements = mapOf(
            "{series}" to ctx.series.replace(WHITESPACE, "_"),
            "{volume}" to ctx.volume.toString(),
            "{page}" to ctx.page.toString(),
            "{date}" to ctx.date,
        )
        return tags
            .map { tag -> VARIABLES.fold(tag) { acc, key -> acc.replace(key, replacements.getValue(key)) } }
            .map { it.trim() }
            .filter { it.isNotEmpty() }
    }
}
